File: android/app/src/main/java/app/ezber/android/models/DisplayOptions.kt
package app.ezber.android.models

/**
 * Per-preset display and playback preferences.
 *
 * [showTransliteration], [showArabic], [showTranslation], [pauseBetweenRepeatsMs]
 * and [playbackMode] map to columns on `presets` in schema/user_schema.sql.
 * [translationMode], [transliterationStyle] and [loopUntilStopped] have no
 * column there, so the Android store keeps them in the `settings` key/value
 * table beside the preset row (see SqliteUserDataStore).
 */
data class DisplayOptions(
    val showTransliteration: Boolean = true,
    val showArabic: Boolean = false,
    val showTranslation: Boolean = true,
    val translationMode: TranslationMode = TranslationMode.TAP_TO_REVEAL,
    val transliterationStyle: TransliterationStyle = TransliterationStyle.STANDARD,
    val pauseBetweenRepeatsMs: Int = 500,
    val loopUntilStopped: Boolean = false,
    val playbackMode: PlaybackMode = PlaybackMode.AYAH,
)

enum class TranslationMode(val label: String) {
    TAP_TO_REVEAL("Tap to reveal"),
    ALWAYS_SHOWN("Always shown"),
    HIDDEN_DURING_PLAYBACK("Hidden during playback"),
    ;

    companion object {
        fun fromName(name: String?): TranslationMode =
            entries.firstOrNull { it.name == name } ?: TAP_TO_REVEAL
    }
}

enum class TransliterationStyle(val label: String) {
    STANDARD("Standard"),
    SIMPLIFIED("Simplified"),
    ;

    companion object {
        fun fromName(name: String?): TransliterationStyle =
            entries.firstOrNull { it.name == name } ?: STANDARD
    }
}

/** Maps to `presets.playback_mode` ('ayah' | 'chapter'). */
enum class PlaybackMode(val schemaValue: String, val label: String) {
    AYAH("ayah", "Verse by verse"),
    CHAPTER("chapter", "Whole chapter"),
    ;

    companion object {
        fun fromSchema(value: String?): PlaybackMode =
            entries.firstOrNull { it.schemaValue == value } ?: AYAH
    }
}

/** Maps to `presets.download_scope` ('inherit' | 'surah' | 'preset'). */
enum class DownloadScope(val schemaValue: String, val label: String) {
    INHERIT("inherit", "Use app default"),
    SURAH("surah", "Whole surah"),
    PRESET("preset", "Preset range only"),
    ;

    companion object {
        fun fromSchema(value: String?): DownloadScope =
            entries.firstOrNull { it.schemaValue == value } ?: INHERIT
    }
}

File: ios/Ezber/Models/DisplayOptions.swift
import Foundation

/// Per-preset display and playback preferences.
struct DisplayOptions: Hashable, Codable {
    var showTransliteration: Bool
    var showArabic: Bool
    var translationMode: TranslationMode
    var transliterationStyle: TransliterationStyle
    var pauseBetweenRepeats: TimeInterval
    var loopUntilStopped: Bool

    init(
        showTransliteration: Bool = true,
        showArabic: Bool = false,
        translationMode: TranslationMode = .tapToReveal,
        transliterationStyle: TransliterationStyle = .standard,
        pauseBetweenRepeats: TimeInterval = 0.5,
        loopUntilStopped: Bool = false
    ) {
        self.showTransliteration = showTransliteration
        self.showArabic = showArabic
        self.translationMode = translationMode
        self.transliterationStyle = transliterationStyle
        self.pauseBetweenRepeats = pauseBetweenRepeats
        self.loopUntilStopped = loopUntilStopped
    }
}

enum TranslationMode: String, Codable, CaseIterable, Identifiable {
    case tapToReveal
    case alwaysShown
    case hiddenDuringPlayback

    var id: String { rawValue }

    var label: String {
        switch self {
        case .tapToReveal: return "Tap to reveal"
        case .alwaysShown: return "Always shown"
        case .hiddenDuringPlayback: return "Hidden during playback"
        }
    }
}

enum TransliterationStyle: String, Codable, CaseIterable, Identifiable {
    case standard
    case simplified

    var id: String { rawValue }

    var label: String {
        switch self {
        case .standard: return "Standard"
        case .simplified: return "Simplified"
        }
    }
}
